**Overall Summary of the Project 2.0**

Dear Kelvin,

Thank you for submitting your second iteration. You've made noticeable progress since your first attempt, filling in missing code cells and improving your data preprocessing. It's clear you’ve put in effort to address previous feedback, and your project now runs more smoothly from data loading through to initial analysis.

***Strengths:***
- **Completed Missing Implementations:**  
  The previously empty code cells are now filled, enabling a full review of each stage.
- **Data Cleaning Accomplishments:**  
  You've renamed columns to snake_case, removed missing values in critical columns, dealt with duplicates, and successfully implemented `replace_wrong_show()` to standardize show types.
- **Filtering and Outlier Identification:**  
  Filtering for shows released in or after 1999 is correctly implemented, and you've identified and attempted to handle outliers in the `imdb_score` distribution.

***Areas of Improvement (Optional):***
- **Code Readability:**  
  Adding more comments to explain the logic behind each code step would further enhance clarity and make it easier for others (and your future self) to understand your reasoning.

***Critical Changes Needed:***
1. **Correct Filtering Condition for Scores:**
   
   Currently, you have:
   <pre><code># filter dataframe using two conditions (scores to be in the range 4-9)
scores = show_df[(show_df['imdb_score']>=4) & (show_df['imdb_votes']<=9)]
</code></pre>

   The second condition uses `imdb_votes` instead of `imdb_score`. This should filter on `imdb_score <= 9` rather than `imdb_votes <= 9`. Using votes instead of scores will give you incorrect subsets of data and skew your final analysis. Correcting this will ensure your data subset accurately represents the intended score range and supports a more valid conclusion about the relationship between show quality and vote counts.

***Final Thoughts:***
Your work has come a long way since the initial submission, and the majority of the pipeline is correctly implemented. With this single but crucial correction to the filtering logic, your project will more accurately reflect the relationship between high ratings and vote counts. Keep up the good work, and after making this fix, we look forward to seeing your final results.

**Status:** **Rejected Pending Completion**

<div style="border:solid blue 2px; padding: 20px">

**Overall Summary of the Project**

Dear Kelvin,

Thank you for submitting your project analyzing the entertainment dataset. Your effort to engage with the data and initiate the preprocessing steps is appreciated. Below is a summarized review of your submission based on the assessment criteria.

---

✅ **Strengths:**

- **Data Loading:**
  - Successfully imported the `pandas` library.
  - Attempted to read the `movies_and_shows.csv` file into a DataFrame, which is the foundational step for data analysis.

- **Initial Exploration:**
  - Made an effort to view the first 10 rows of the dataset using `df.head(10)`, a good practice for understanding the data structure.

---

**Missing Code Implementations:**
   
   - **Issue:** Several code cells contain only comments without the corresponding code implementation. This prevents the analysis from being executed and reviewed effectively.
    
---

🌱 **Final Thoughts**

Dear Kelvin,

Your project demonstrates a commendable start with data loading and initial exploration. However, to fully assess and validate your analysis, it's essential to complete all the required steps in the data preprocessing and analysis stages. Completing the missing code cells will allow for a comprehensive review of your work and ensure that your findings accurately reflect the data.

**Recommendation:** Please complete all sections of your project, ensuring that each step is properly implemented. Once done, feel free to resubmit your project for a thorough review.

**Status:** **Rejected Pending Completion**

</div>

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [25]:
import pandas as pd


Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [26]:
df= pd.read_csv('/datasets/movies_and_shows.csv')


Print the first 10 table rows:

In [27]:
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [29]:
df= pd.read_csv('/datasets/movies_and_shows.csv')

df = df.rename(columns={
    'Character': 'character',
    'r0le': 'role',
    'TITLE': 'title',
    '  Type':'type',
    'release Year': 'release_year',
    'imdb sc0re': 'imdb_score',
    'imdb v0tes': 'imdb_votes'
})

Check the result. Print the names of the columns once more:

In [30]:
df.head(10)

name                      character   role        title   type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release_year              genres  imdb_score  imdb_votes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [31]:
# calculating missing values

print(df.isna().sum())

   name            0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64


We identified missing values in several columns. While the missing value in `'title'` isn't critical, missing values in `'imdb_score'` and `'imdb_votes'` affect around 6% of the data, which could impact our analysis. To ensure data integrity, we'll drop all rows with missing values.

In [32]:
# dropping rows where columns with scores, and votes have missing values
# 1. assign the operation to a varible
# 2. don't drop the whole column, using the `axis=0`
df = df.dropna(subset=['imdb_score', 'imdb_votes'], axis=0)


In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80853 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       80853 non-null  object 
 1   character     80853 non-null  object 
 2   role          80853 non-null  object 
 3   title         80853 non-null  object 
 4   type          80853 non-null  object 
 5   release_year  80853 non-null  int64  
 6   genres        80853 non-null  object 
 7   imdb_score    80853 non-null  float64
 8   imdb_votes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 6.2+ MB


After doing this, I dropped the missing values, then I checked the # rows, I found the row number is 80853. Comparing to the orignal row number 85578. The diff of them is 4725 

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [34]:
# counting missing values
df.isna().sum()



   name         0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [35]:
df.duplicated().sum()



6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [36]:
filter = df.duplicated(keep=False)
df[filter]

name                  character      role  \
7560         Philip Greene  Baseball Fan (uncredited)     ACTOR   
7561         Philip Greene  Baseball Fan (uncredited)     ACTOR   
14502             Dan Levy                   Reporter     ACTOR   
14512             Dan Levy                   Reporter     ACTOR   
18951     Nicolas Le Nev??                    unknown  DIRECTOR   
...                    ...                        ...       ...   
85569       Jessica Cediel            Liliana Navarro     ACTOR   
85570  Javier Gardeaz?­bal   Agust??n "Peluca" Ort??z     ACTOR   
85571        Carla Giraldo             Valery Reinoso     ACTOR   
85572  Ana Mar??a S?­nchez                    Lourdes     ACTOR   
85577         Isabel Gaona                     Cacica     ACTOR   

                                 title   type  release_year  \
7560                   How Do You Know  MOVIE          2010   
7561                   How Do You Know  MOVIE          2010   
14502  A Very Harold & Kumar Christmas  MOVIE          2011   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18951                       Sammy & Co   SHOW          2014   
...                                ...    ...           ...   
85569                          Lokillo  MOVIE          2021   
85570                          Lokillo  MOVIE          2021   
85571                          Lokillo  MOVIE          2021   
85572                          Lokillo  MOVIE          2021   
85577                          Lokillo  MOVIE          2021   

                                 genres  imdb_score  imdb_votes  
7560     ['comedy', 'drama', 'romance']         5.4     50383.0  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14502  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18951         ['animation', 'european']         5.7        31.0  
...                                 ...         ...         ...  
85569                        ['comedy']         3.8        68.0  
85570                        ['comedy']         3.8        68.0  
85571                        ['comedy']         3.8        68.0  
85572                        ['comedy']         3.8        68.0  
85577                        ['comedy']         3.8        68.0  

[13988 rows x 9 columns]

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [37]:
# removing duplicate rows
df = df.drop_duplicates()


Check for duplicate rows once more to make sure you have removed all of them:

In [38]:
# checking for duplicates
df.duplicated().sum()

0

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [39]:
# viewing unique type names
unique_type = df['type'].sort_values().unique() 
print(unique_type)

['MOVIE' 'SHOW' 'movies' 'shows' 'the movie' 'tv' 'tv series' 'tv show'
 'tv shows']


Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [40]:
# function for replacing implicit duplicates 
def replace_wrong_show(wrong_show_list, correct_show):
    
    df['type'] = df['type'].replace(wrong_show_list, correct_show)    

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [41]:
# removing implicit duplicates
wrong_show_list =["shows", "SHOW", "tv show", "tv shows", "tv series", "tv"]
correct_show ="show"
replace_wrong_show(wrong_show_list, correct_show)

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [42]:
# viewing unique genre names
df['type'].sort_values().unique()

array(['MOVIE', 'movies', 'show', 'the movie'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [43]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
golden = df['type'] == 'show'
ages = df['release_year'] >= 1999
show_df = df[golden]
show_after_1999_df = show_df[ages]
show_after_1999_df

/tmp/ipykernel_32/2670528704.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  show_after_1999_df = show_df[ages]


name                character      role      title  \
1664           Jeff Probst           Himself - Host     ACTOR   Survivor   
2076         Mayumi Tanaka  Monkey D. Luffy (voice)     ACTOR  One Piece   
2077          Kazuya Nakai     Roronoa Zoro (voice)     ACTOR  One Piece   
2078         Akemi Okamura             Nami (voice)     ACTOR  One Piece   
2079      Kappei Yamaguchi            Usopp (voice)     ACTOR  One Piece   
...                    ...                      ...       ...        ...   
85433     Maneerat Kam-Uan                       Ae     ACTOR  Let's Eat   
85434    Rudklao Amratisha                  unknown     ACTOR  Let's Eat   
85435      Jaturong Mokjok                  unknown     ACTOR  Let's Eat   
85436     Pisamai Wilaisak                  unknown     ACTOR  Let's Eat   
85437  Sarawut Wichiensarn                  unknown  DIRECTOR  Let's Eat   

       type  release_year                                             genres  \
1664   show          2000                                        ['reality']   
2076   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2077   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2078   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2079   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
...     ...           ...                                                ...   
85433  show          2021                                ['drama', 'comedy']   
85434  show          2021                                ['drama', 'comedy']   
85435  show          2021                                ['drama', 'comedy']   
85436  show          2021                                ['drama', 'comedy']   
85437  show          2021                                ['drama', 'comedy']   

       imdb_score  imdb_votes  
1664          7.4     24687.0  
2076          8.8    117129.0  
2077          8.8    117129.0  
2078          8.8    117129.0  
2079          8.8    117129.0  
...           ...         ...  
85433         8.2         5.0  
85434         8.2         5.0  
85435         8.2         5.0  
85436         8.2         5.0  
85437         8.2         5.0  

[13430 rows x 9 columns]

In [44]:
# repeat conditional indexing so df has only shows (movies are removed as result)
show_df = df[(df['type']== 'show') & (df['release_year'] >= 1999)]
show_df

name                character      role      title  \
1664           Jeff Probst           Himself - Host     ACTOR   Survivor   
2076         Mayumi Tanaka  Monkey D. Luffy (voice)     ACTOR  One Piece   
2077          Kazuya Nakai     Roronoa Zoro (voice)     ACTOR  One Piece   
2078         Akemi Okamura             Nami (voice)     ACTOR  One Piece   
2079      Kappei Yamaguchi            Usopp (voice)     ACTOR  One Piece   
...                    ...                      ...       ...        ...   
85433     Maneerat Kam-Uan                       Ae     ACTOR  Let's Eat   
85434    Rudklao Amratisha                  unknown     ACTOR  Let's Eat   
85435      Jaturong Mokjok                  unknown     ACTOR  Let's Eat   
85436     Pisamai Wilaisak                  unknown     ACTOR  Let's Eat   
85437  Sarawut Wichiensarn                  unknown  DIRECTOR  Let's Eat   

       type  release_year                                             genres  \
1664   show          2000                                        ['reality']   
2076   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2077   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2078   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2079   show          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
...     ...           ...                                                ...   
85433  show          2021                                ['drama', 'comedy']   
85434  show          2021                                ['drama', 'comedy']   
85435  show          2021                                ['drama', 'comedy']   
85436  show          2021                                ['drama', 'comedy']   
85437  show          2021                                ['drama', 'comedy']   

       imdb_score  imdb_votes  
1664          7.4     24687.0  
2076          8.8    117129.0  
2077          8.8    117129.0  
2078          8.8    117129.0  
2079          8.8    117129.0  
...           ...         ...  
85433         8.2         5.0  
85434         8.2         5.0  
85435         8.2         5.0  
85436         8.2         5.0  
85437         8.2         5.0  

[13430 rows x 9 columns]

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [45]:
# rounding column with scores

#checking the outcome with tail()

show_df['imdb_score']= show_df['imdb_score'].round(0)

show_df.tail()

/tmp/ipykernel_32/2636726969.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  show_df['imdb_score']= show_df['imdb_score'].round(0)


name character      role      title  type  release_year  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  show          2021   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  show          2021   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  show          2021   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  show          2021   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  show          2021   

                    genres  imdb_score  imdb_votes  
85433  ['drama', 'comedy']         8.0         5.0  
85434  ['drama', 'comedy']         8.0         5.0  
85435  ['drama', 'comedy']         8.0         5.0  
85436  ['drama', 'comedy']         8.0         5.0  
85437  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [46]:
# Use groupby() for scores and count all unique values in each group, print the result
show_df.groupby('imdb_score')['imdb_votes'].count().reset_index()


imdb_score  imdb_votes
0         2.0          24
1         3.0          27
2         4.0         180
3         5.0         592
4         6.0        2494
5         7.0        4706
6         8.0        4842
7         9.0         557
8        10.0           8

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [47]:
# filter dataframe using two conditions (scores to be in the range 4-9)
scores=show_df[(show_df['imdb_score']>=4) & (show_df['imdb_score']<=9)]
# group scores and corresponding average number of votes, reset index and print the result
avg_votes = scores.groupby('imdb_score')['imdb_votes'].mean().reset_index()
avg_votes
                                            

imdb_score     imdb_votes
0         4.0    5277.583333
1         5.0    3143.942568
2         6.0    3481.717322
3         7.0    8727.068211
4         8.0   30299.460967
5         9.0  126904.109515

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [48]:
# round column with averages
avg_votes['imdb_votes']= avg_votes['imdb_votes'].round(0)

# rename columns
avg_votes = avg_votes.rename(columns={'imdb_votes':'imdb_averages'})

# print dataframe in descending order
avg_votes.sort_values(by='imdb_averages', ascending=False)

imdb_score  imdb_averages
5         9.0       126904.0
4         8.0        30299.0
3         7.0         8727.0
0         4.0         5278.0
2         6.0         3482.0
1         5.0         3144.0

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.